src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -505,6 +505,298 @@ public TextReader CreateTextReader()
+        {
+            if (this.readingStream)
+            {
+                throw JsonReaderExtensions.CreateException(Strings.JsonReader_CannotCallReadInStreamState);
+            }
+
+            if (this.canStream)
+            {
+                this.canStream = false;
+                if (this.nodeType == JsonNodeType.PrimitiveValue)
+                {
+                    // caller is positioned on a string value that they haven't read, so skip it
+                    if (this.characterBuffer[this.tokenStartIndex] == 'n')
+                    {
+                        await this.ParseNullPrimitiveValueAsync().ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        await this.ParseStringPrimitiveValueAsync().ConfigureAwait(false);
+                    }
+                }
+            }
+
+            // Reset the node value.
+            this.nodeValue = null;
+
+#if DEBUG
+            // Reset the node type to None - so that we can verify that the Read method actually sets it.
+            this.nodeType = JsonNodeType.None;
+#endif
+
+            // Skip any whitespace characters.
+            // This also makes sure that we have at least one non-whitespace character available.
+            if (!await this.SkipWhitespacesAsync().ConfigureAwait(false))
+            {
+                return this.EndOfInput();
+            }
+
+            Debug.Assert(
+                this.tokenStartIndex < this.storedCharacterCount && !IsWhitespaceCharacter(this.characterBuffer[this.tokenStartIndex]),
+                "The SkipWhitespacesAsync didn't correctly skip all whitespace characters from the input.");
+
+            Scope currentScope = this.scopes.Peek();
+
+            bool commaFound = false;
+            if (this.characterBuffer[this.tokenStartIndex] == ',')
+            {
+                commaFound = true;
+                this.tokenStartIndex++;
+
+                // Note that validity of the comma is verified below depending on the current scope.
+                // Skip all whitespaces after comma.
+                // Note that this causes "Unexpected EOF" error if the comma is the last thing in the input.
+                // It might not be the best error message in certain cases, but it's still correct (a JSON payload can never end in comma).
+                if (!await this.SkipWhitespacesAsync().ConfigureAwait(false))
+                {
+                    return this.EndOfInput();
+                }
+
+                Debug.Assert(
+                    this.tokenStartIndex < this.storedCharacterCount && !IsWhitespaceCharacter(this.characterBuffer[this.tokenStartIndex]),
+                    "The SkipWhitespacesAsync didn't correctly skip all whitespace characters from the input.");
+            }
+
+            switch (currentScope.Type)
+            {
+                case ScopeType.Root:
+                    if (commaFound)

```

Comment:
I almost feel like this `switch` would be easier to understand if there was one for `commaFound == true` and one for `commaFound == false`. Probably not worth changing now, but maybe something to consider.